Extended Model64 Support

gamemasterplc opened the pull request:
Now supports hierarchical and skinned models. Also implemented support for glCopyMatrixN64 to copy a matrix from a matrix stack to the current matrix stack and made gldemo use the function.

gamemasterplc committed:
Implement hierarchical models on n64 side

gamemasterplc committed:
Add basic skinning data for model64

Also changed type of node references in models to an integer.

gamemasterplc committed:
Allow null mesh/skin pointers

gamemasterplc committed:
Implement extended GLTF model tool

The tool and runtime is untested

gamemasterplc committed:
Fix model64_set_node_rot

gamemasterplc committed:
Fix some bugs with nodes in mkmodel

gamemasterplc committed:
Force 4-Byte Alignment for Indices

gamemasterplc committed:
Fix node matrix for node that is parent of scene

gamemasterplc committed:
Implement glCopyMatrixN64

gamemasterplc committed:
Generate scalar matrix index buffer

gamemasterplc committed:
Correct models with matrix index buffers

gamemasterplc committed:
Use weights to improve matrix index buffer

gamemasterplc committed:
Backup RA in GLCmd_MatrixCopy

gamemasterplc committed:
Fix RSP Crash for GLCmd_MatrixCopy

gamemasterplc committed:
Do cpu copy as well in glCopyMatrixN64

gamemasterplc committed:
Change skinned model example to use glCopyMatrixN64

gamemasterplc committed:
Fix stride of vertex attributes

gamemasterplc committed:
Fix unksinned models

gamemasterplc committed:
Add inverse bind matrices to model64

gamemasterplc committed:
Fix Inverse Bind Matrix Accessor

gamemasterplc committed:
Fix matrix calculation for skinned models

gamemasterplc committed:
Fix documentation build

gamemasterplc committed:
Add model64_clone to public API

snacchus commented on include/model64.h:
```diff
@@ -30,6 +34,47 @@ uint32_t model64_get_mesh_count(model64_t *model);
+/**
+ * @brief Return the node at the specified index.
+ */
+model64_node_t *model64_get_node(model64_t *model, uint32_t mesh_index);
```
```suggestion
model64_node_t *model64_get_node(model64_t *model, uint32_t node_index);
```

snacchus commented on src/GL/rsp_gl.S:
```diff
@@ -348,6 +350,27 @@ GLCmd_MatrixLoad:
+    #       Bit 1..0: Source Matrix
+    #############################################################
+    move ra2, ra
+    andi a0, a0, 0x3
```
This line is not needed. The command ID is in bits 31..24 (and even gets shifted upwards in the next line), so it's ignored when using it as a DMEM address
```suggestion
```

snacchus commented on src/GL/rsp_gl.S:
```diff
@@ -348,6 +350,27 @@ GLCmd_MatrixLoad:
+    #############################################################
+    move ra2, ra
+    andi a0, a0, 0x3
+    sll rhs, a0, 6
```
Instead of shifting here you could pre-shift on the CPU and save this opcode.

snacchus commented on src/model64.c:
```diff
@@ -10,88 +11,376 @@
+    return sizeof(model64_t)+(model_data->num_nodes*sizeof(node_transform_state_t));
+}
+
+static void multiply_node_mtx(float parent[16], float child[16])
+{
+    
+    for (int i = 0; i < 4; ++i)
+    {
+        float l0 = child[i * 4 + 0];
+        float l1 = child[i * 4 + 1];
+        float l2 = child[i * 4 + 2];
+
+        float r0 = l0 * parent[0] + l1 * parent[4] + l2 * parent[8];
+        float r1 = l0 * parent[1] + l1 * parent[5] + l2 * parent[9];
+        float r2 = l0 * parent[2] + l1 * parent[6] + l2 * parent[10];
+
+        child[i * 4 + 0] = r0;
+        child[i * 4 + 1] = r1;
+        child[i * 4 + 2] = r2;
+    }
+
+    child[12] += parent[12];
+    child[13] += parent[13];
+    child[14] += parent[14];
+}
```
If I understand correctly, this an optimized matrix multiplication which assumes that both matrices are translation+rotation+scale only?

gamemasterplc replied:
I think so, its copied from cgltf.

snacchus replied:
I think I get what it does now. It's basically just a shortened matrix multiplication that assumes the last row is (0, 0, 0, 1). Looks fine to me.

snacchus commented on src/model64.c:
```diff
@@ -10,88 +11,376 @@
+    for(int i=0; i<16; i++)
+    {
+        dst[i] = src[i];
+    }
```
Could also just use memcpy instead

snacchus commented on src/model64.c:
```diff
@@ -10,88 +11,376 @@
+    child[14] += parent[14];
+}
+
+static void mtx_copy(float src[16], float dst[16])
```
Maybe this is just me, but I'm used to the convention of the first parameter being the destination and the second the source, like in memcpy for example

snacchus commented on src/model64_internal.h:
```diff
@@ -34,21 +34,73 @@ typedef struct primitive_s {
+    mesh_t *mesh;                   ///< Mesh a node refers to
+    model64_skin_t *skin;           ///< Skin a node refers to
+    node_transform_t transform;     ///< Initial transform of a node
+    uint32_t parent;                ///< Index of parent node
```
I was wondering why the parent is represented with an index instead of a direct pointer. I assume the reason is that keeping pointers would make loading/unloading the structure more complicated? Or is there another reason?

gamemasterplc replied:
It makes it easier to look up node matrices for an instance if one uses an index rather than a pointer.

snacchus commented on src/model64.c:
```diff
@@ -10,88 +11,376 @@
+    dst[1] = y;
+    dst[2] = z;
+    calc_node_local_matrix(model, node_idx);
+    update_node_matrices(model);
```
Would it not be sufficient to only update `node` and its descendants? You could also keep intermediate results and multiply the matrices iteratively.

gamemasterplc replied:
That was removed while I was trying to get node world matrices to work correctly.

snacchus replied:
Ok, we can always add optimizations later on

snacchus commented on src/model64.c:
```diff
@@ -10,88 +11,376 @@
+    dst[2] = z;
+    dst[3] = w;
+    calc_node_local_matrix(model, node_idx);
+    update_node_matrices(model);
```
See comment in `model64_set_node_pos`

snacchus commented on src/model64.c:
```diff
@@ -10,88 +11,376 @@
+    dst[1] = y;
+    dst[2] = z;
+    calc_node_local_matrix(model, node_idx);
+    update_node_matrices(model);
```
See comment in `model64_set_node_pos`

snacchus commented on src/model64.c:
```diff
@@ -10,88 +11,376 @@
+
+static uint32_t get_node_idx(model64_t *model, model64_node_t *node)
+{
+    return node-model->data->nodes;
```
Maybe just personal preference, but I'd prefer spaces around the minus operator for better readability

snacchus commented on src/model64.c:
```diff
@@ -10,88 +11,376 @@
+void model64_get_node_world_mtx(model64_t *model, model64_node_t *node, float dst[16])
+{
+    uint32_t node_idx = get_node_idx(model, node);
+    assertf(node_idx < model->data->num_nodes, "Setting scale of invalid node.");
```
Copy paste error in the assertion message

snacchus commented on src/model64.c:
```diff
@@ -155,9 +444,42 @@ void model64_draw_mesh(mesh_t *mesh)
+            {
+                glCurrentPaletteMatrixARB(i);
+                glCopyMatrixN64(GL_MODELVIEW); //Copy matrix at top of modelview stack to matrix palette
+                glMultMatrixf(model->transforms[node->skin->joints[i].node_idx].world_mtx);
```
Does this also work if the skin's root node is not actually a root node in the entire hierarchy? I suppose it does, since it just applies the transforms of all ancestors as well, which is expected. 
In glTF, a skin also specifies its root node: https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#_skin_skeleton
Is that not needed in our case? I'm not sure what it would actually be required for.

gamemasterplc replied:
The skeleton can be safely ignored in glTF. But I'm not sure if multiplying the modelview matrix by the node's world matrix is correct for skinned models.

snacchus replied:
Does it still work if you try to transform the entire model via GL_MODELVIEW?

gamemasterplc replied:
It works if you change the modelview before calling model64_draw_node.

snacchus commented on src/model64_internal.h:
```diff
@@ -34,21 +34,73 @@ typedef struct primitive_s {
+
+typedef struct node_transform_state_s {
+    node_transform_t transform;     ///< Current transform state for a node
+    float world_mtx[16];            ///< World matrix for a node
```
Is the naming "world_mtx" actually accurate? I guess it's in "world space" relative to the scene that is described by the model file. But one could also interpret that to mean world space in the application as a whole (even though that doesn't make much sense). Maybe "model_mtx" instead to avoid confusion?

gamemasterplc replied:
Would that also change the name of model64_get_node_world_mtx.

snacchus replied:
It would, but now I'm second guessing my proposed naming. Maybe "world_mtx" is fine for now.

snacchus reviewed (changes requested):
I've looked at all files except mkmodel.c because I've run out of time for now.

Looks good to me overall. The only thing I'm unsure about is whether skins should also specify their skeleton's root node like in glTF (see my comment).

Also I think updating the joint matrices could be optimized a lot more by working iteratively, instead of re-traversing the entire hierarchy again for each node.

rasky commented on include/GL/gl.h:
```diff
@@ -321,6 +321,8 @@ void glPopMatrix(void);
 
 void glCurrentPaletteMatrixARB(GLint index);
 
+void glCopyMatrixN64(GLenum source);
```
Don't we need to declare an extension for this? Is this part of another extension?

gamemasterplc replied:
Can I just add it to the extensions string in query.c? I don't think its part of another extension.

snacchus replied:
Yes, you need to add it to the extensions string, and also add a macro at the top of gl.h (all the other extensions are there as well). You can probably call the extension `GL_N64_copy_matrix` or something.

rasky commented on src/GL/rsp_gl.S:
```diff
@@ -348,6 +350,27 @@ GLCmd_MatrixLoad:
 
+GLCmd_MatrixCopy:
+    #############################################################
+    # GLCmd_SetFlag
```
This name is wrong in the comment

rasky commented on src/model64.c:
```diff
@@ -10,88 +11,376 @@
+    float cp = cosf(y * 0.5);
+    float sp = sinf(y * 0.5);
+    float cy = cosf(z * 0.5);
+    float sy = sinf(z * 0.5);
```
please use `fm_sincosf` here

gamemasterplc replied:
I don't have fm_sincosf in my branch of libdragon.

rasky commented on tools/mkmodel/mkmodel.c:
```diff
@@ -581,6 +832,120 @@ int convert(const char *infn, const char *outfn)
+            if (cgltf_accessor_unpack_floats(ibm_accessor, ibm_buffer, num_values) == 0) {
+                fprintf(stderr, "Error: failed reading attribute data\n");
+                free(ibm_buffer);
+                return 1;
```
Shouldn't this be a goto error as well?

gamemasterplc replied:
Yes, it should.

gamemasterplc committed:
Implement some fixes

Changed second parameter name of model64_get_node to node_index.
Improve GLCmd_MatrixCopy and fix comments in rsp_gl.s.
Change mtx_copy to use memcpy.
Fix model64_get_node_world_mtx assert message.
Cleanup inverse bind matrix access failure in mkmodel.c.

snacchus commented on tools/mkmodel/mkmodel.c:
```diff
@@ -552,13 +791,25 @@ int convert(const char *infn, const char *outfn)
         goto error;
     }
-
+    if (data->nodes_count <= 0) {
+        fprintf(stderr, "Error: input file contains no nodes\n");
+        goto error;
+    }
+    if (!data->scene) {
+        fprintf(stderr, "Error: input file has no default scene\n");
+        goto error;
+    }
+    if (data->scene->nodes_count <= 0) {
+        fprintf(stderr, "Error: input file has no root node in default scene\n");
+        goto error;
+    }
```
Do we have to error out here? What if someone just wants to import some meshes and handle everything else (transform, animation etc.) themselves? Obviously skinned meshes will need nodes, but skins can only be defined on nodes anyway.

gamemasterplc replied:
I changed it not to error out.

snacchus commented on tools/mkmodel/mkmodel.c:
```diff
@@ -581,6 +832,120 @@ int convert(const char *infn, const char *outfn)
     }
+    // Convert skins
+    model->num_skins = data->skins_count;
+    model->skins = calloc(data->skins_count, sizeof(model64_skin_t));
```
Probably shouldn't allocate if there are no skins

snacchus commented on tools/mkmodel/mkmodel.c:
```diff
@@ -581,6 +832,120 @@ int convert(const char *infn, const char *outfn)
+    if(data->scene->nodes_count > 1) {
+        model->num_nodes++;
+    }
+    model->nodes = calloc(model->num_nodes, sizeof(model64_node_t));
```
Probably shouldn't allocate if there are no nodes (I know that would error out atm anyway, but see my other comment on that)

snacchus commented on tools/mkmodel/mkmodel.c:
```diff
@@ -150,85 +221,171 @@ void indices_write(FILE *out, uint32_t type, void *data, uint32_t count)
-            attribute_write(out, &primitive->texcoord,  &vertices_placeholders[cur_primitive*ATTRIBUTE_COUNT + 2]);
-            attribute_write(out, &primitive->normal,    &vertices_placeholders[cur_primitive*ATTRIBUTE_COUNT + 3]);
-            attribute_write(out, &primitive->mtx_index, &vertices_placeholders[cur_primitive*ATTRIBUTE_COUNT + 4]);
+            uint32_t color_data_ofs = data_ofs+attribute_get_data_size(&primitive->position);
+            uint32_t texcoord_data_ofs = color_data_ofs+attribute_get_data_size(&primitive->color);
+            uint32_t normal_data_ofs = texcoord_data_ofs+attribute_get_data_size(&primitive->texcoord);
+            uint32_t mtx_index_data_ofs = normal_data_ofs+attribute_get_data_size(&primitive->normal);
+            uint32_t index_ofs = mtx_index_data_ofs+attribute_get_data_size(&primitive->mtx_index);
```
I get that the placeholders might be a bit cumbersome to use and I assume that's why you moved the code away from them. But I do think that they make the code easier to reason about. Now, whenever something about the file layout changes, you always have to make sure that the precomputed offsets are still correct, otherwise you'll get bugs that are hard to track. That problem does not exist with the placeholder technique.

We could try and come up with something that would make the placeholders less verbose maybe.

gamemasterplc replied:
I think that would involve something with a hash map that records a pointer, where the pointer points to in the file, and a list of references. The hashmap would index by pointer. We can use the stb data structures to help implement it.

gamemasterplc committed:
Add GL_N64_copy_matrix to list of extensions

Also clean up some mkmodel code.
